### PR TITLE
Various tweaks and changes for discussion

### PR DESCRIPTION
**Pull request by daviddrysdale**
These are in a variety of areas, so probably easiest to review commit-by-commit

### CODE REVIEW

File: server/trillian_log_server.go [hunk 1]
```diff
@@ -18,17 +19,47 @@ import (
...
+	c.mu.Lock()
+	defer c.mu.Unlock()
+
+	s, ok := c.logStorages[logID]
+	if ok {
+		return s, nil
+	}
+
+	s, err := c.provider(logID)
+	if err == nil {
```
Comment (AlCutter) [inline server/trillian_log_server.go]: Maybe flip this and return err early, and then store + return below?
(indent error flow)

```go
s, err := ...
if err != nil {
   return nil, err
}
c.logStorages[logID] = s
return s, nil
```

File: server/vmap/trillian_map_server.go [hunk 2]
```diff
@@ -53,7 +53,7 @@ func (t *TrillianMapServer) getStorageForMap(mapID int64) (storage.MapStorage, e
 
 	s, err := t.storageProvider(mapID)
 
-	if err != nil {
+	if err == nil {
```
Comment (AlCutter) [inline server/vmap/trillian_map_server.go]: Ah, yes, it's all _my_ fault :)

File: log/sequencer.go [hunk 3]
```diff
@@ -11,13 +11,15 @@ import (
...
+	"golang.org/x/net/context"
 )
 
-// Sequencer instances are responsible for integrating new leaves into a log.
+// Sequencer instances are responsible for integrating new leaves into a single log.
 // Leaves will be assigned unique sequence numbers when they are processed.
 // There is no strong ordering guarantee but in general entries will be processed
 // in order of submission to the log.
 type Sequencer struct {
+	ctx        context.Context
```
Comment (AlCutter) [inline log/sequencer.go]: Feels a bit weird to store this here, would it make sense to pass it in as a parameter to `SequenceBatch()` and pass it around?

File: server/trillian_log_server.go (hunk 1 above)
Comment (daviddrysdale) [inline server/trillian_log_server.go]: Done.

File: log/sequencer.go (hunk 3 above)
Comment (daviddrysdale) [inline log/sequencer.go]: Done, although the net result makes me wonder whether a better approach might be to drop the `Context` key:value and instead add a `MapID()` method to the various log storage `interface`s, analogous to `MapStorage.MapID()`.  Then the diagnostics could just get the ID back from the storage instance.  Preferences?

**Comment by daviddrysdale** [comment]
PTAL

File: log/sequencer.go [hunk 4]
```diff
@@ -110,20 +111,20 @@ func (s Sequencer) sequenceLeaves(mt *merkle.CompactMerkleTree, leaves []trillia
...
+	return s.buildMerkleTreeFromStorageAtRoot(ctx, currentRoot, tx)
 }
 
-func (s Sequencer) signRoot(root trillian.SignedLogRoot) (trillian.DigitallySigned, error) {
+func (s Sequencer) createRootSignature(ctx context.Context, root trillian.SignedLogRoot) (trillian.DigitallySigned, error) {
 	signer, err := s.keyManager.Signer()
 
 	if err != nil {
-		glog.Warningf("key manager failed to create crypto.Signer: %v", err)
+		glog.Warningf("%skey manager failed to create crypto.Signer: %v", util.LogIDPrefix(ctx), err)
```
Comment (AlCutter) [inline log/sequencer.go]: Having the %sjoined with the following message is causing me a horrible mental itch :)
Could we squander some bytes and have it be "%s: …"?

File: server/log_rpc_server.go [hunk 5]
```diff
@@ -17,21 +19,53 @@ import (
...
 
-// LogStorageProviderFunc decouples the server from storage implementations
+// LogStorageProviderFunc generates a LogStorage instance for a given log ID; it is a
+// separate type in order to decouple the server from storage implementations.
 type LogStorageProviderFunc func(int64) (storage.LogStorage, error)
 
+// cachedLogStorageProvider wraps a LogStorageProviderFunc and caches its results.
+type cachedLogStorageProvider struct {
+	provider    LogStorageProviderFunc
+	mu          sync.Mutex
```
Comment (AlCutter) [inline server/log_rpc_server.go]: Maybe pop a blank line above `mu` to show the relatedness?

File: server/log_rpc_server.go [hunk 6]
```diff
@@ -17,21 +19,53 @@ import (
...
+	logStorages map[int64]storage.LogStorage // guarded by mu
+}
+
+// newCachedLogStorageProvider creates a new cached storage provider for the given
+// raw provider.
+func newCachedLogStorageProvider(p LogStorageProviderFunc) cachedLogStorageProvider {
+	return cachedLogStorageProvider{provider: p, logStorages: make(map[int64]storage.LogStorage)}
```
Comment (AlCutter) [inline server/log_rpc_server.go]: Since these members are named, move them down onto a line each?

File: server/vmap/trillian_map_server/main.go [hunk 7]
```diff
@@ -121,6 +115,16 @@ func main() {
 		glog.Fatalf("Failed to load map server key: %v", err)
 	}
 
+	// Start HTTP server (optional)
+	if *exportRPCMetrics {
+		err := startHTTPServer(*httpPortFlag)
+
+		if err != nil {
+			glog.Fatalf("Failed to start http server on port %d: %v", *httpPortFlag, err)
+			os.Exit(1)
```
Comment (AlCutter) [inline server/vmap/trillian_map_server/main.go]: The `Fatalf` will terminate, so the call to `os.Exit` can go I think.

File: util/context_test.go [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+import (
+	"testing"
+
+	"golang.org/x/net/context"
+)
+
+func TestLogContext(t *testing.T) {
+	ctx := context.Background()
+	want := "{unknown} "
+	if got := LogIDPrefix(ctx); got != want {
```
Comment (AlCutter) [inline util/context_test.go]: Would it make sense to make these table-driven?

File: log/sequencer.go (hunk 4 above)
Comment (daviddrysdale) [inline log/sequencer.go]: Done

File: server/log_rpc_server.go (hunk 5 above)
Comment (daviddrysdale) [inline server/log_rpc_server.go]: Done

File: server/log_rpc_server.go (hunk 6 above)
Comment (daviddrysdale) [inline server/log_rpc_server.go]: Done

File: server/vmap/trillian_map_server/main.go (hunk 7 above)
Comment (daviddrysdale) [inline server/vmap/trillian_map_server/main.go]: Done (and in the log server)

File: util/context_test.go (hunk 8 above)
Comment (daviddrysdale) [inline util/context_test.go]: Done

**Pull request merged**